design/fake-client.rst

```diff
@@ -0,0 +1,153 @@
+
+The first design option is probably less effort, but it has the drawback that
+subtle changes in the API may be forgotten. One example for that is the
+recent introduction of an optimization for findall() by name, which now uses
+a cached mapping of names to URIs, but had the subtle externally visible
+change that the resource objects returned by findall(name) now only have
+a very minimal set of properties, and no longer those that are returned by
+list() which was the earlier implementation. For example, the "status"
+property is no longer set in that case. The second design would automatically
+represent this example of a change.
+
+Design
+------
+
+In the following description, the second design option is used.
+
+There is a fake session object that the user of the zhmcclient package
+will use instead of zhmcclient.Session. The fake session object contains a
+faked HMC that can represent all resources relevant for the zhmcclient.
+There fake session object implements the same API as zhmcclient.Session, so
+that the rest of the zhmcclient classes does not need to be changed.
+The get(), post() and delete() methods of the fake session class will
+be redirected to operate on the faked HMC.
+
+The fake session object will have the ability to be "loaded" with an
+initial state of resources. The resource definitions can be provided as
+a dictionary of nested dictionaries, following this example::
+
+    resources = {
+        'cpcs': [  # name of manager attribute for this resource
+            {
+                'properties': {
+                    # object-id and object-uri are auto-generated
+                    'name': 'cpc_1',
+                    . . .  # more properties
+                },
+                'adapters': [
+                    {
+                        'properties': {
+                            # object-id and object-uri are auto-generated
+                            'name': 'ad_1',
+                            . . .  # more properties
+                        },
+                        'ports': [
+                            {
+                                'properties': {
+                                    # element-id and element-uri are
+                                    # auto-generated
+                                    'name': 'port_1',
+                                    . . .  # more properties
+                                }
+                            },
+                            . . .  # more Ports
+                        ],
+                    },
+                    . . .  # more Adapters
+                ],
+                . . .  # more CPC child resources
+            },
+            . . .  # more CPCs
+        ]
+    }
+
+The properties for object ID ("object-id" or "element-id") and URI
+("object-uri" or "element-uri") are auto-generated, if not provided by
+the user. All other properties will only exist as provided by the user.
+There is no checking for invalid properties.
```

Comment:
Checking for invalid properties would require to know the names of all properties f all resource types. Right now, the zhmcclient code and the prototype code for this fake client have to know only a very small number of properties (object-id, object-uri, name). Plus, the set of properties depends on the API version, and probably some even depend on the CPC generation. Trying to avoid that knowledge about all properties was really the reason for this statement.